Use the highest QOS available when detecting content with DataDetectors

rr-codes opened the pull request:
#### 0d7c0048ca22e42c1c974a3a832562d8fec781e0
<pre>
Use the highest QOS available when detecting content with DataDetectors
<a href="https://bugs.webkit.org/show_bug.cgi?id=267879">https://bugs.webkit.org/show_bug.cgi?id=267879</a>
<a href="https://rdar.apple.com/118585870">rdar://118585870</a>

Reviewed by Aditya Keerthi.

Use the highest QOS available when creating a DDScanner.

To facilitate this, make the content data detection mechanism async.

* Source/WebCore/PAL/pal/spi/cocoa/DataDetectorsCoreSPI.h:
* Source/WebCore/editing/cocoa/DataDetection.h:
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::buildQuery):
(WebCore::workQueue):
(WebCore::parseAllResultRanges):
(WebCore::DDQueryFragmentCore::operator== const):
(WebCore::getFragmentsFromQuery):
(WebCore::processDataDetectorScannerResults):
(WebCore::DataDetection::detectContentInFrame):
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebKit/Shared/Cocoa/DataDetectionResult.h:
* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInRangeHandle.mm:
(-[WKWebProcessPlugInRangeHandle detectDataWithTypes:context:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::detectDataInFrame):
(WebKit::WebPage::detectDataInAllFrames):

Canonical link: <a href="https://commits.webkit.org/274699@main">https://commits.webkit.org/274699@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e7454edf527f9c99a38ef7d933990dc06afe1a4c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39796 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18807 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42340 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35698 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42103 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21701 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16137 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40370 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15847 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/34415 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13716 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13735 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35370 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43619 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36137 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12021 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16230 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16278 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5235 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/editing/cocoa/DataDetection.h:
```diff
@@ -31,6 +31,7 @@
 #import "FloatRect.h"
 #import "SimpleRange.h"
 #import <wtf/OptionSet.h>
+
```
undo

rr-codes commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -659,6 +698,72 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+
+    auto contextRange = makeRangeSelectingNodeContents(*document);
+
+    auto scanner = adoptCF(PAL::softLink_DataDetectorsCore_DDScannerCreate(DDScannerTypeStandard, 0, nullptr));
```
actually set the property

rr-codes commented on Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInRangeHandle.mm:
```diff
@@ -70,7 +70,7 @@ - (NSString *)text
 {
 #if ENABLE(DATA_DETECTION)
-    return WebCore::DataDetection::detectContentInRange(makeSimpleRange(_rangeHandle->coreRange()), fromWKDataDetectorTypes(types), WebCore::DataDetection::extractReferenceDate(context));
+    return nil;
```
put this back

rr-codes commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -659,6 +698,72 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+    if (types.contains(DataDetectorType::LookupSuggestion))
+        PAL::softLink_DataDetectorsCore_DDScannerEnableOptionalSource(scanner.get(), DDScannerSourceSpotlight, true);
+
+    workQueue().dispatch([scanner = WTFMove(scanner), types, referenceDateFromContext, scanQuery = WTFMove(scanQuery), frame, fragments = WTFMove(fragments), completionHandler = WTFMove(completionHandler)]() mutable {
```
use frame ids

pxlcoder commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2785,16 +2787,41 @@ void FrameLoader::checkLoadCompleteForThisFrame()
+                    if (m_frame->isMainFrame() && m_frame->page()) {
+                        ASSERT(&m_frame->page()->mainFrame() == m_frame.ptr());
+                        protectedFrame()->protectedPage()->diagnosticLoggingClient().logDiagnosticMessageWithResult(DiagnosticLoggingKeys::pageLoadedKey(), emptyString(), DiagnosticLoggingResultPass, ShouldSample::Yes);
+                    }
```
What does this frame loading code have to do with DataDetection?

rr-codes replied:
Because originally, this code was being executed post-detection; so I wanted to maintain that order

pxlcoder replied:
It would be good to understand if that is really necessary. If not, we wouldn't be taking full advantage of the async nature of the code, and potentially slowing down other things with the higher QOS.

rr-codes replied:
it works without, so doesn’t seem needed

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -449,26 +451,113 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+    return queue.get();
+}
 
-    // FIXME: we should add a timeout to this call to make sure it doesn't take too much time.
```
Is this FIXME no longer relevant? It seems like the FIXME was simply removed.

rr-codes replied:
yes, no longer relevant I don’t think since it is no longer blocking the main thread

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -449,26 +451,113 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
     if (!CFArrayGetCount(scannerResults.get()))
         return nil;
 
+    auto tempScanQuery = adoptCF(PAL::softLink_DataDetectorsCore_DDScanQueryCreate(NULL));
```
`RetainPtr` instead of `auto`, per explicit usage guidelines.

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -449,26 +451,113 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+
+static NSArray * processDataDetectorScannerResults(DDScannerRef scanner, OptionSet<DataDetectorType> types, std::optional<double> referenceDateFromContext, DDScanQueryRef scanQuery, const SimpleRange& contextRange, const Vector<DDQueryFragmentCore>& oldFragments)
+{
+    auto scannerResults = adoptCF(PAL::softLink_DataDetectorsCore_DDScannerCopyResultsWithOptions(scanner, PAL::get_DataDetectorsCore_DDScannerCopyResultsOptionsForPassiveUse() | DDScannerCopyResultsOptionsCoalesceSignatures));
```
`RetainPtr` instead of `auto`, per explicit usage guidelines.

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.h:
```diff
@@ -63,7 +69,8 @@ class DataDetection {
     WEBCORE_EXPORT static std::optional<DetectedItem> detectItemAroundHitTestResult(const HitTestResult&);
 #endif
-    WEBCORE_EXPORT static NSArray *detectContentInRange(const SimpleRange&, OptionSet<DataDetectorType>, std::optional<double> referenceDate);
+    WEBCORE_EXPORT static void detectContentInFrame(const WeakPtr<LocalFrame>&, OptionSet<DataDetectorType>, std::optional<double>, CompletionHandler<void(NSArray *)>&&);
```
I think it's fine for parameters to use raw pointers, but let's see if @cdumez has something to say.

cdumez replied:
Yes, fine to pass parameters as raw pointers. Of course, this assumes that the call site has a strong ref to this pointer on the stack before passing it in argument.

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -656,13 +695,87 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+
+    auto contextRange = makeRangeSelectingNodeContents(*document);
+
+    auto scanner = adoptCF(PAL::softLink_DataDetectorsCore_DDScannerCreate(DDScannerTypeStandard, 0, nullptr));
```
`RetainPtr`

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -656,13 +695,87 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+#if HAVE(DDSCANNER_QOS_CONFIGURATION)
+    PAL::softLink_DataDetectorsCore_DDScannerSetQOS(scanner.get(), DDQOSHighest);
+#endif
+    auto scanQuery = adoptCF(PAL::softLink_DataDetectorsCore_DDScanQueryCreate(NULL));
```
`RetainPtr`

pxlcoder commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -656,13 +695,87 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+NSArray *DataDetection::detectContentInRange(const SimpleRange& contextRange, OptionSet<DataDetectorType> types, std::optional<double> referenceDateFromContext)
+{
+    auto scanner = adoptCF(PAL::softLink_DataDetectorsCore_DDScannerCreate(DDScannerTypeStandard, 0, nullptr));
+    auto scanQuery = adoptCF(PAL::softLink_DataDetectorsCore_DDScanQueryCreate(NULL));
```
`RetainPtr`

pxlcoder commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2791,10 +2791,15 @@ void FrameLoader::checkLoadCompleteForThisFrame()
+                        return;
+
+                    auto documentLevelResults = retainPtr(results);
+                    m_frame->dataDetectionResults().setDocumentLevelResults(documentLevelResults.get());
```
You need to protect `m_frame` somewhere, because this method could destroy it, and it's used below.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4721,22 +4721,34 @@ void WebPage::removeDataDetectedLinks(CompletionHandler<void(const DataDetection
+
+    DataDetection::detectContentInFrame(WeakPtr { localFrame.get() }, dataDetectorTypes, dataDetectionReferenceDate, [localFrame, mainFrameResult = WTFMove(mainFrameResult), dataDetectionReferenceDate, completionHandler = WTFMove(completionHandler), dataDetectorTypes](NSArray *results) mutable {
+        auto retainedResults = retainPtr(results);
+        localFrame->dataDetectionResults().setDocumentLevelResults(retainedResults.get());
```
localFrame should be protected.

cdumez commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -647,7 +686,7 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
 
             lastNodeContent = currentTextNode->data().substring(range.end.offset, currentTextNode->length() - range.end.offset);
             lastTextNodeToUpdate = WTFMove(currentTextNode);
-        }        
```
Would be really nice if you could fix your IDE settings to stop doing this. So much noise in the PR because of this.

cdumez commented on Source/WebCore/editing/cocoa/DataDetection.mm:
```diff
@@ -656,13 +695,87 @@ static inline CFComparisonResult queryOffsetCompare(DDQueryOffset o1, DDQueryOff
+            return;
+        }
+
+        WTF::callOnMainRunLoop([scanner = WTFMove(scanner), types, referenceDateFromContext, scanQuery = WTFMove(scanQuery), frame, fragments = WTFMove(fragments), completionHandler = WTFMove(completionHandler)]() mutable {
```
Do we really need `WTF::` here? I doubt it.

cdumez commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2791,10 +2791,15 @@ void FrameLoader::checkLoadCompleteForThisFrame()
-                if (m_frame->isMainFrame())
-                    m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
+                DataDetection::detectContentInFrame(m_frame, types, m_client->dataDetectionReferenceDate(), [this, weakFrame = WeakPtr { m_frame.get() }, document](NSArray *results) {
+                    if (!weakFrame || &weakFrame->loader() != this)
```
*FIRST* thing you should do before using weakFrame is store it is a RefPtr and then only use that RefPtr from now on.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4721,22 +4721,34 @@ void WebPage::removeDataDetectedLinks(CompletionHandler<void(const DataDetection
-    completionHandler(WTFMove(mainFrameResult));
+
+    DataDetection::detectContentInFrame(WeakPtr { localFrame.get() }, dataDetectorTypes, dataDetectionReferenceDate, [localFrame, mainFrameResult = WTFMove(mainFrameResult), dataDetectionReferenceDate, completionHandler = WTFMove(completionHandler), dataDetectorTypes](NSArray *results) mutable {
+        auto retainedResults = retainPtr(results);
```
This is fishy. The caller of this lambda should already have retained results before passing it to us. This seems to go against our smart pointer guidelines.

rr-codes replied:
this is existing code that I just moved, so I did not want to risk changing it

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4721,22 +4721,34 @@ void WebPage::removeDataDetectedLinks(CompletionHandler<void(const DataDetection
+        if (localFrame->isMainFrame())
+            mainFrameResult->results = retainedResults;
+
+        auto next = localFrame->tree().traverseNext();
```
Can we use `Ref` here directly?

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4721,22 +4721,34 @@ void WebPage::removeDataDetectedLinks(CompletionHandler<void(const DataDetection
+            return;
+        }
+
+        detectDataInFrame(Ref { *next }, dataDetectorTypes, dataDetectionReferenceDate, WTFMove(mainFrameResult), WTFMove(completionHandler));
```
Then we don't need `Ref { }` here.

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -4721,22 +4721,34 @@ void WebPage::removeDataDetectedLinks(CompletionHandler<void(const DataDetection
+void WebPage::detectDataInAllFrames(OptionSet<WebCore::DataDetectorType> dataDetectorTypes, CompletionHandler<void(const DataDetectionResult&)>&& completionHandler)
+{
+    auto mainFrameResult = makeUniqueRef<DataDetectionResult>();
+    detectDataInFrame(Ref { m_page->mainFrame() }, dataDetectorTypes, m_dataDetectionReferenceDate, WTFMove(mainFrameResult), WTFMove(completionHandler));
```
You can simply call `m_page->protectedMainFrame()`, no need for `Ref { }`

pxlcoder commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2791,10 +2791,16 @@ void FrameLoader::checkLoadCompleteForThisFrame()
-                if (m_frame->isMainFrame())
-                    m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
+                DataDetection::detectContentInFrame(protectedFrame().ptr(), types, m_client->dataDetectionReferenceDate(), [this, weakFrame = WeakPtr { m_frame.get() }, document](NSArray *results) {
+                    RefPtr strongFrame { weakFrame.get() };
```
`protectedFrame` or just `frame`.

rr-codes committed:
Use the highest QOS available when detecting content with DataDetectors
https://bugs.webkit.org/show_bug.cgi?id=267879
rdar://118585870

Reviewed by Aditya Keerthi.

Use the highest QOS available when creating a DDScanner.

To facilitate this, make the content data detection mechanism async.

* Source/WebCore/PAL/pal/spi/cocoa/DataDetectorsCoreSPI.h:
* Source/WebCore/editing/cocoa/DataDetection.h:
* Source/WebCore/editing/cocoa/DataDetection.mm:
(WebCore::buildQuery):
(WebCore::workQueue):
(WebCore::parseAllResultRanges):
(WebCore::DDQueryFragmentCore::operator== const):
(WebCore::getFragmentsFromQuery):
(WebCore::processDataDetectorScannerResults):
(WebCore::DataDetection::detectContentInFrame):
(WebCore::DataDetection::detectContentInRange):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
* Source/WebKit/Shared/Cocoa/DataDetectionResult.h:
* Source/WebKit/WebProcess/InjectedBundle/API/Cocoa/WKWebProcessPlugInRangeHandle.mm:
(-[WKWebProcessPlugInRangeHandle detectDataWithTypes:context:]):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::detectDataInFrame):
(WebKit::WebPage::detectDataInAllFrames):

Canonical link: https://commits.webkit.org/274699@main
